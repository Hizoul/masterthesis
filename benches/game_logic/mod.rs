pub mod tetromino;
pub mod field;
pub mod transposition;