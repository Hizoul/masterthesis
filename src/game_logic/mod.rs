pub mod val;
pub mod cache;
pub mod tetromino;
pub mod log;
pub mod field;