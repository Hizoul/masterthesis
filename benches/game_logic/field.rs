extern crate criterion;

use criterion::Criterion;
use rustyblocks::game_logic::field::{GameField};
use rustyblocks::game_logic::field::helper_structs::{GameRules};
use rustyblocks::game_logic::log::{GameLog};
use rustyblocks::game_logic::cache::{ShapeCache};
use std::sync::{Arc};
use rand::prelude::*;

const BLA: &'static str = "{\"log\":[{\"PayloadRolled\":{\"from\":1,\"block\":3}},{\"PayloadPlaced\":{\"from\":1,\"block\":3,\"orientation\":0,\"x\":8,\"y\":0}},{\"PayloadRolled\":{\"from\":0,\"block\":3}},{\"PayloadPlaced\":{\"from\":0,\"block\":3,\"orientation\":0,\"x\":6,\"y\":0}},{\"PayloadRolled\":{\"from\":1,\"block\":3}},{\"PayloadPlaced\":{\"from\":1,\"block\":3,\"orientation\":0,\"x\":4,\"y\":0}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":0,\"x\":7,\"y\":2}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":6,\"y\":2}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":7,\"y\":3}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":6,\"y\":6}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":1,\"x\":6,\"y\":10}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":2,\"y\":0}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":1,\"x\":6,\"y\":14}},{\"PayloadRolled\":{\"from\":1,\"block\":4}},{\"PayloadPlaced\":{\"from\":1,\"block\":4,\"orientation\":1,\"x\":7,\"y\":5}},{\"PayloadRolled\":{\"from\":0,\"block\":4}},{\"PayloadPlaced\":{\"from\":0,\"block\":4,\"orientation\":5,\"x\":9,\"y\":3}},{\"PayloadRolled\":{\"from\":1,\"block\":4}},{\"PayloadPlaced\":{\"from\":1,\"block\":4,\"orientation\":3,\"x\":8,\"y\":6}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":1,\"x\":10,\"y\":6}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":2,\"y\":2}},{\"PayloadRolled\":{\"from\":0,\"block\":4}},{\"PayloadPlaced\":{\"from\":0,\"block\":4,\"orientation\":5,\"x\":8,\"y\":7}},{\"PayloadRolled\":{\"from\":1,\"block\":2}},{\"PayloadPlaced\":{\"from\":1,\"block\":2,\"orientation\":2,\"x\":4,\"y\":2}},{\"PayloadRolled\":{\"from\":0,\"block\":3}},{\"PayloadPlaced\":{\"from\":0,\"block\":3,\"orientation\":1,\"x\":7,\"y\":10}},{\"PayloadRolled\":{\"from\":1,\"block\":2}},{\"PayloadPlaced\":{\"from\":1,\"block\":2,\"orientation\":0,\"x\":4,\"y\":3}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":1,\"x\":2,\"y\":6}},{\"PayloadRolled\":{\"from\":1,\"block\":2}},{\"PayloadPlaced\":{\"from\":1,\"block\":2,\"orientation\":3,\"x\":5,\"y\":5}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":9,\"y\":10}},{\"PayloadRolled\":{\"from\":1,\"block\":3}},{\"PayloadPlaced\":{\"from\":1,\"block\":3,\"orientation\":3,\"x\":4,\"y\":6}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadPlaced\":{\"from\":0,\"block\":2,\"orientation\":3,\"x\":8,\"y\":11}},{\"PayloadRolled\":{\"from\":1,\"block\":2}},{\"PayloadPlaced\":{\"from\":1,\"block\":2,\"orientation\":1,\"x\":4,\"y\":9}},{\"PayloadRolled\":{\"from\":0,\"block\":3}},{\"PayloadPlaced\":{\"from\":0,\"block\":3,\"orientation\":3,\"x\":7,\"y\":12}},{\"PayloadRolled\":{\"from\":1,\"block\":4}},{\"PayloadPlaced\":{\"from\":1,\"block\":4,\"orientation\":1,\"x\":0,\"y\":0}},{\"PayloadRolled\":{\"from\":0,\"block\":3}},{\"PayloadPlaced\":{\"from\":0,\"block\":3,\"orientation\":1,\"x\":4,\"y\":12}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":9,\"y\":12}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":9,\"y\":14}},{\"PayloadRolled\":{\"from\":1,\"block\":4}},{\"PayloadPlaced\":{\"from\":1,\"block\":4,\"orientation\":3,\"x\":0,\"y\":3}},{\"PayloadRolled\":{\"from\":0,\"block\":4}},{\"PayloadPlaced\":{\"from\":0,\"block\":4,\"orientation\":5,\"x\":0,\"y\":4}},{\"PayloadRolled\":{\"from\":1,\"block\":3}},{\"PayloadPlaced\":{\"from\":1,\"block\":3,\"orientation\":1,\"x\":4,\"y\":14}},{\"PayloadRolled\":{\"from\":0,\"block\":4}},{\"PayloadPlaced\":{\"from\":0,\"block\":4,\"orientation\":7,\"x\":0,\"y\":5}},{\"PayloadRolled\":{\"from\":1,\"block\":4}},{\"PayloadPlaced\":{\"from\":1,\"block\":4,\"orientation\":7,\"x\":1,\"y\":8}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":9,\"y\":16}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":3,\"y\":4}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadPlaced\":{\"from\":0,\"block\":2,\"orientation\":3,\"x\":5,\"y\":15}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":3,\"y\":8}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":9,\"y\":18}},{\"PayloadRolled\":{\"from\":1,\"block\":2}},{\"PayloadPlaced\":{\"from\":1,\"block\":2,\"orientation\":1,\"x\":2,\"y\":12}},{\"PayloadRolled\":{\"from\":0,\"block\":3}},{\"PayloadPlaced\":{\"from\":0,\"block\":3,\"orientation\":1,\"x\":2,\"y\":15}},{\"PayloadSkipped\":{\"from\":1,\"block\":4,\"reason\":1}},{\"PayloadRolled\":{\"from\":0,\"block\":4}},{\"PayloadPlaced\":{\"from\":0,\"block\":4,\"orientation\":5,\"x\":7,\"y\":15}},{\"PayloadSkipped\":{\"from\":1,\"block\":0,\"reason\":1}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadPlaced\":{\"from\":0,\"block\":2,\"orientation\":2,\"x\":3,\"y\":16}},{\"PayloadRolled\":{\"from\":1,\"block\":3}},{\"PayloadPlaced\":{\"from\":1,\"block\":3,\"orientation\":3,\"x\":0,\"y\":10}},{\"PayloadSkipped\":{\"from\":0,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":1,\"reason\":1}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":2,\"y\":17}},{\"PayloadSkipped\":{\"from\":0,\"block\":4,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":4,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":4,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":4,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":2,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":4,\"reason\":1}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":4,\"y\":17}},{\"PayloadSkipped\":{\"from\":0,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":1,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":2,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":1,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":2,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":1,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":0,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":0,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":4,\"reason\":1}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":2,\"y\":19}},{\"PayloadSkipped\":{\"from\":0,\"block\":1,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":3,\"reason\":1}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadPlaced\":{\"from\":0,\"block\":2,\"orientation\":2,\"x\":7,\"y\":18}},{\"PayloadSkipped\":{\"from\":1,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":1,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":0,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":0,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":4,\"reason\":1}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadConsidering\":{\"play_index\":0}}]}";


struct Foo {
  counter: usize
}

impl Foo {
  pub fn add(&mut self) {self.counter+=1;}
  pub fn get(&self) -> usize {self.counter}
}

trait FooTrait {
  fn t_add(&mut self);
  fn t_get(&self) -> usize;
}

impl FooTrait for Foo {
  fn t_add(&mut self) {self.add();}
  fn t_get(&self) -> usize {self.get()}
}

pub fn bench_field(c: &mut Criterion) {
  let cache = ShapeCache::new();
  let rules = GameRules::default();
  let p = Arc::new(cache);
  let p2 = p.clone();
  let p3 = p.clone();
  let log: GameLog = serde_json::from_str(BLA).unwrap();
  c.bench_function("deterministic_play_until_over", move |b| b.iter(|| {
    let mut f = GameField::new_with_cache(2, p.clone(), rules);
    while !f.game_over {
      f.place_block_using_play(0);
    }
  }));
  c.bench_function("random_play_until_over", move |b| b.iter(|| {
    let mut  f = GameField::new_with_cache(2, p2.clone(), rules);
    while !f.game_over {
      let p = f.possible_plays.len() as u8;
      f.place_block_using_play(rand::thread_rng().gen_range(0,p));
    }
  }));
  c.bench_function("replay_log", move |b| b.iter(|| {
    let mut f = GameField::new_with_cache(2, p3.clone(), rules);
    f.restore_from_log(&log, false);
  }));
  c.bench_function("traitcost_direct", move |b| b.iter(|| {
    let mut f = Foo {counter:0};
    for _ in 0..99999999 {
      while f.get() < 9999999999 {
        f.add();
      }
    }
  }));
  c.bench_function("traitcost_struct", move |b| b.iter(|| {
    let mut f: Box<dyn FooTrait> = Box::new(Foo {counter:0});
    for _ in 0..99999999 {
      while f.t_get() < 9999999999 {
        f.t_add();
      }
    }
  }));
}