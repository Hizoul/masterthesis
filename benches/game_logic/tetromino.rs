extern crate criterion;

use criterion::Criterion;
use rustyblocks::game_logic::tetromino::{Position,get_block_shape,get_translated_shape};
use rand::prelude::*;

pub fn bench_base_shape(c: &mut Criterion) {
  c.bench_function("block_shape", |b| b.iter(||
    get_block_shape(
      rand::thread_rng().gen_range(0,5),
      rand::thread_rng().gen_range(0,4)
    )));
}
pub fn bench_translated_shape(c: &mut Criterion) {
  c.bench_function("translate_shape", |b| b.iter(|| {
    let mut rng = rand::thread_rng();
    get_translated_shape(
      rng.gen_range(0,5),
      rng.gen_range(0,4),
      &Position::new(rng.gen_range(0,10),rng.gen_range(0,20))
    )
  }));
}