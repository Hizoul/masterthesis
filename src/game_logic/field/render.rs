use super::GameField;
use crate::game_logic::val::{GAMEFIELD_MAX_WIDTH,GAMEFIELD_MAX_WIDTH_USIZE,GAMEFIELD_MAX_HEIGHT_USIZE,GAMEFIELD_MAX_HEIGHT,AMOUNT_OF_UNIQUE_PIECES_USIZE};

pub trait Renderable {
  fn to_field_string(&self) -> String;
  fn to_field_string_with_player(&self, include_current_player: bool) -> String;
  fn to_field_string_with_player_as_u8(&self) -> Vec<u8>;
  fn to_playable_string(&mut self) -> String;
  fn to_reshapeable_array(&mut self, second_player_as_first: bool) -> Vec<i8>;
  fn to_reshaped_array(&mut self) -> Vec<Vec<i8>>;
  fn to_image(&mut self, second_player_as_first: bool) -> Vec<f32>;
  fn converted(&mut self) -> Vec<f32>;
}
pub const MAX: usize = GAMEFIELD_MAX_HEIGHT_USIZE*GAMEFIELD_MAX_WIDTH_USIZE;
pub const MAX_FOR_SCORE: usize = (GAMEFIELD_MAX_HEIGHT_USIZE+1)*GAMEFIELD_MAX_WIDTH_USIZE;
pub const MAX_FOR_BLOCKS: usize = (GAMEFIELD_MAX_HEIGHT_USIZE+3)*GAMEFIELD_MAX_WIDTH_USIZE;
// 1 for scores, 2 and 3 for block amounts of players, 19 more for valid play indicators
pub const AMOUNT_OF_ADDITIONAL_LINES: usize = 22;

impl Renderable for GameField {
  fn to_field_string(&self) -> String {
    self.to_field_string_with_player(false)
  }
  fn to_field_string_with_player(&self, include_current_player: bool) -> String {
    let mut field = String::with_capacity((GAMEFIELD_MAX_WIDTH_USIZE * GAMEFIELD_MAX_HEIGHT_USIZE) +
        GAMEFIELD_MAX_WIDTH_USIZE + 1);
    field.push('\n');
    for ys in 0..GAMEFIELD_MAX_HEIGHT {
      let y = (GAMEFIELD_MAX_HEIGHT - 1) - ys;
      for x in 0..GAMEFIELD_MAX_WIDTH {
        let mut player = "0".to_string();
        'search: for p in self.tetrominos.iter() {
          let mut i = 0;
          for part in self.shape_cache.get_block_shape(p).iter() {
            if part.x == x && part.y == y {
              let to_set = if i == 0 {((p.from + 1) * 2) - 1} else {((p.from + 1) * 2)};
              player = to_set.to_string();
              break 'search;
            }
            i += 1;
          }
        }
        field.push_str(player.as_str());
      }
      field.push('\n');
    }
    if include_current_player {
      field.push_str(self.current_player.to_string().as_str());
    }
    field
  }
  fn to_field_string_with_player_as_u8(&self) -> Vec<u8> {
    let mut field: [u8; 201] = [0; 201];
    for p in self.tetrominos.iter() {
      let mut i = 0;
      for part in self.shape_cache.get_block_shape(p).iter() {
        let to_set = if i == 0 {i+=1;((p.from + 1) * 2) - 1} else {((p.from + 1) * 2)};
        field[((((GAMEFIELD_MAX_HEIGHT_USIZE-1) - part.y as usize) * GAMEFIELD_MAX_WIDTH_USIZE) + part.x as usize)] = to_set as u8;
      }
    }
    field[200] = self.current_player as u8;
    field.to_vec()
  }
  fn to_reshapeable_array(&mut self, second_player_as_first: bool) -> Vec<i8> {
    // first additional line = scores
    // second and third additional lines = blocks per players
    // 190 additional entries for possible combinations
    let mut field_nums: Vec<i8> = Vec::with_capacity((GAMEFIELD_MAX_HEIGHT_USIZE+AMOUNT_OF_ADDITIONAL_LINES) * GAMEFIELD_MAX_WIDTH_USIZE);
    for index in 0..(GAMEFIELD_MAX_HEIGHT_USIZE+AMOUNT_OF_ADDITIONAL_LINES)*GAMEFIELD_MAX_WIDTH_USIZE {
      if index >= MAX {
        if index < MAX_FOR_SCORE {
          let mut corrected_index = index % MAX;
          if corrected_index < (self.players as usize) {
            if second_player_as_first {
              corrected_index = if corrected_index == 0 {1} else {0};
            }
            field_nums.push(self.scores[corrected_index] as i8);
          } else {
            field_nums.push(-1);
          }
        } else if index < MAX_FOR_BLOCKS {
          let mut player = (index % MAX_FOR_SCORE) / AMOUNT_OF_UNIQUE_PIECES_USIZE;
          if player >= (self.players as usize) {
            field_nums.push(-1);
          } else {
            let block = (index % MAX_FOR_SCORE) % AMOUNT_OF_UNIQUE_PIECES_USIZE;
            if second_player_as_first {
              player = if player == 0 {1} else {0};
            }
            field_nums.push(self.piece_counter[player][block]);
          }
        } else {
          field_nums.push(-1);
        }
      } else {
        field_nums.push(0);
      }
    }
    for play in self.possible_plays.iter() {
      field_nums[(MAX_FOR_BLOCKS)+play.global_index as usize] = 1;
    }
    for p in self.tetrominos.iter() {
      let mut i = 0;
      for part in self.shape_cache.get_block_shape(p).iter() {
        let mut from = p.from;
        if second_player_as_first {
          from = if from == 0 {1} else {0};
        }
        let to_set = if i == 0 {
          i += 1;
          ((from + 1) * 2) - 1
        } else {
          ((from + 1) * 2)
        };
        field_nums[((((GAMEFIELD_MAX_HEIGHT_USIZE-1) - part.y as usize) * GAMEFIELD_MAX_WIDTH_USIZE) + part.x as usize)] = to_set;
      }
    }
    // let b = self.current_block;
    // let o = self.current_orientation;
    // let x = self.current_x;
    // let y = self.highest_y(b, o, x);
    // if !self.game_over {
    //   let hover_block = self.shape_cache.get_translated(b, o, x, y);
    //   for part in hover_block.iter() {
    //     field_nums[((((GAMEFIELD_MAX_HEIGHT_USIZE-1) - part.y as usize) * GAMEFIELD_MAX_WIDTH_USIZE) + part.x as usize)] = (self.current_player as u8 + 1) * 2;
    //   }
    // }
    field_nums
  }
  fn to_image(&mut self, second_player_as_first: bool) -> Vec<f32> {
    let mut to_return: Vec<f32> = vec![0.0; 3072];
    let orig = self.to_reshapeable_array(second_player_as_first);
    for i in 0..200 {
      let y = (i as f32 / 10.0).floor() as usize;
      let x = i % 10;
      let mut current = 0;
      let mut val = 0.0;
      if orig[i] == 1 || orig[i] == 2 {
        current = 0;
        val = if orig[i] == 1 {128.0} else {255.0};
      } else if orig[i] == 3 || orig[i] == 4 {
        current = 2;
        val = if orig[i] == 3 {128.0} else {255.0};
      }
      to_return[((y * 32*3) + x*3) + current] = val;
    }
    for i in 200..210 {
      let y = ((i - 200) as f32 / 10.0).floor() as usize;
      let x = (i % 10) + 10;
      to_return[(y*32*3)+x*3+1] = orig[i] as f32 + 124.0;
    }
    for i in 210..230 {
      let y = ((i - 200) as f32 / 10.0).floor() as usize;
      let x = (i % 10) + 10;
      let val = if orig[i] <= 0 {0.0} else {orig[i] as f32 * 50.0};
      to_return[(y*32*3)+x*3+1] = val;
    }
    for i in 230..420 {
      let y = ((i - 200) as f32 / 10.0).floor() as usize;
      let x = (i % 10) + 10;
      let val = if orig[i] == -1 {
         0.0
      } else {
        255.0
      };
      to_return[(y*32*3)+x*3+1] = val;
    }
    to_return
  }
  fn converted(&mut self) -> Vec<f32> {
    let not_yet_reshaped = self.to_reshapeable_array(false);
    let mut converted: Vec<f32> = Vec::new();
    for num in not_yet_reshaped {
      converted.push(num as f32);
    }
    converted
  }
  fn to_reshaped_array(&mut self) -> Vec<Vec<i8>> {
    let not_yet_reshaped = self.to_reshapeable_array(false);
    let mut reshaped = Vec::new();
    let mut current_sub = Vec::new();
    let i = 0;
    for num in not_yet_reshaped {
      current_sub.push(num);
      if i > 0 && i % 10 == 0 {
        reshaped.push(current_sub.clone());
        current_sub.clear();
      }
    }
    reshaped
  }
  fn to_playable_string(&mut self) -> String {
    let mut field_nums: Vec<Vec<i8>> = Vec::with_capacity(GAMEFIELD_MAX_HEIGHT_USIZE);
    for _ in 0..GAMEFIELD_MAX_HEIGHT {
      let mut new_line = Vec::with_capacity(GAMEFIELD_MAX_WIDTH_USIZE);
      for _ in 0..GAMEFIELD_MAX_WIDTH {
        new_line.push(0);
      }
      field_nums.push(new_line);
    }
    for p in self.tetrominos.iter() {
      for part in self.shape_cache.get_block_shape(p).iter() {
        field_nums[(GAMEFIELD_MAX_HEIGHT_USIZE-1) - part.y as usize][part.x as usize] = ((p.from + 1) * 2) - 1;
      }
    }
    let b = self.current_block;
    let o = self.current_orientation;
    let x = self.current_x;
    let y = self.highest_y(b, o, x);
    if !self.game_over {
      let hover_block = self.shape_cache.get_translated(b, o, x, y);
      for part in hover_block.iter() {
        let new_y = (GAMEFIELD_MAX_HEIGHT_USIZE-1) - part.y as usize;
        let new_x = part.x as usize;
        if new_x < GAMEFIELD_MAX_WIDTH_USIZE && new_y < GAMEFIELD_MAX_HEIGHT_USIZE {
          field_nums[new_y][new_x] = (self.current_player + 1) * 2;
        }
      }
    }
    let mut field = String::with_capacity((GAMEFIELD_MAX_WIDTH_USIZE * GAMEFIELD_MAX_HEIGHT_USIZE) +
        GAMEFIELD_MAX_WIDTH_USIZE + 1);
    for top in field_nums.iter() {
      for ele in top.iter() {
        field.push_str(ele.to_string().as_str());
      }
      field.push('\n');
    }
    field
  }
}