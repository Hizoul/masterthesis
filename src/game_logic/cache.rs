use super::val::{GAMEFIELD_MAX_HEIGHT,GAMEFIELD_MAX_WIDTH,AMOUNT_OF_UNIQUE_PIECES_USIZE,AMOUNT_OF_UNIQUE_ORIENTATIONS_USIZE,GAMEFIELD_MAX_WIDTH_USIZE, GAMEFIELD_MAX_HEIGHT_USIZE,
AMOUNT_OF_UNIQUE_PIECES,AMOUNT_ROTATIONS};
use super::tetromino::{Tetromino,Position,BlockShape,get_translated_shape,get_block_shape};

#[derive(Debug,Clone)]
pub struct ShapeCache {
  base: Vec<Vec<BlockShape>>,
  translated: Vec<Vec<Vec<Vec<BlockShape>>>>
}

impl ShapeCache {
  pub fn new() -> ShapeCache {
    let mut base = Vec::with_capacity(AMOUNT_OF_UNIQUE_PIECES_USIZE);
    let mut translated = Vec::with_capacity(AMOUNT_OF_UNIQUE_PIECES_USIZE);
    for block in 0..AMOUNT_OF_UNIQUE_PIECES {
      let mut rotations = Vec::with_capacity(AMOUNT_OF_UNIQUE_ORIENTATIONS_USIZE);
      let mut translated_rotations = Vec::with_capacity(AMOUNT_OF_UNIQUE_ORIENTATIONS_USIZE);
      for rotation in 0..AMOUNT_ROTATIONS[block as usize] {
        rotations.push(get_block_shape(block, rotation));
        let mut xs = Vec::with_capacity(GAMEFIELD_MAX_WIDTH_USIZE+1);
        for x in 0..(GAMEFIELD_MAX_WIDTH) {
          let mut ys = Vec::with_capacity(GAMEFIELD_MAX_HEIGHT_USIZE+1);
          for y in 0..(GAMEFIELD_MAX_HEIGHT) {
            ys.push(get_translated_shape(block, rotation, &Position {x,y}));
          }
          xs.push(ys);
        }
        translated_rotations.push(xs)
      }
      base.push(rotations);
      translated.push(translated_rotations);
    }
    return ShapeCache {
      base, translated
    }
  }
  pub fn get_translated(&self, block: i8, rotation:i8, x: i8, y: i8) -> &BlockShape {
    self.translated.get(block as usize).expect("block available").get(rotation as usize).expect("rotation available").get(x as usize).expect("x available").get(y as usize).expect("y available")
  }
  pub fn get_base(&self, block: i8, rotation:i8) -> &BlockShape {
    self.base.get(block as usize).unwrap().get(rotation as usize).unwrap()
  }
  pub fn get_block_shape(&self, block: &Tetromino) -> &BlockShape {
    self.get_translated(block.block, block.rotation, block.position.x, block.position.y)
  }
}
