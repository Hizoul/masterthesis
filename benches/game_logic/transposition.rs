extern crate criterion;

use criterion::Criterion;
use rustyblocks::game_logic::field::{GameField};
use rustyblocks::game_logic::field::helper_structs::{GameRules};
use rustyblocks::game_logic::field::render::{Renderable};
use rustyblocks::zobrist::{ZobristHasher,U64Hasher};
use std::collections::HashMap;

pub fn bench_transpo(c: &mut Criterion) {
  c.bench_function("hash_u8", move |b| b.iter(|| {
    let mut set = HashMap::new();
    for _ in 0..3 {
      let mut f = GameField::new_with_rules(2, GameRules::deterministic());
      while !f.game_over {
        f.place_block_using_play(0);
        set.insert(f.to_field_string_with_player_as_u8(), 1);
      }
    }
  }));
  c.bench_function("hash_string", move |b| b.iter(|| {
    let mut set = HashMap::new();
    for _ in 0..3 {
      let mut f = GameField::new_with_rules(2, GameRules::deterministic());
      while !f.game_over {
        f.place_block_using_play(0);
        set.insert(f.to_field_string_with_player(true), 1);
      }
    }
  }));
  let mut zobrist_hasher2 = ZobristHasher::new(201, 20);
  c.bench_function("hash_zobrist_unopt", move |b| b.iter(|| {
    let hash_builder = U64Hasher::new();
    let mut set = HashMap::with_hasher(hash_builder);
    for _ in 0..3 {
      let mut f = GameField::new_with_rules(2, GameRules::deterministic());
      while !f.game_over {
        f.place_block_using_play(0);
        set.insert(zobrist_hasher2.hash(&f.to_field_string_with_player_as_u8()), 1);
      }
    }
  }));
  let mut zobrist_hasher3 = ZobristHasher::new(201, 20);
  c.bench_function("hash_zobrist_unopt_nou64hasher", move |b| b.iter(|| {
    let mut set = HashMap::new();
    for _ in 0..3 {
      let mut f = GameField::new_with_rules(2, GameRules::deterministic());
      while !f.game_over {
        f.place_block_using_play(0);
        set.insert(zobrist_hasher3.hash(&f.to_field_string_with_player_as_u8()), 1);
      }
    }
  }));
  let mut zobrist_hasher = ZobristHasher::new(201, 20);
  c.bench_function("hash_zobrist_field", move |b| b.iter(|| {
    let mut set = HashMap::new();
    for _ in 0..3 {
      let mut f = GameField::new_with_rules(2, GameRules::deterministic());
      while !f.game_over {
        f.place_block_using_play(0);
        set.insert(zobrist_hasher.hash_field(&f), 1);
      }
    }
  }));
}