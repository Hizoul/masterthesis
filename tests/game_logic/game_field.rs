use rustyblocks::game_logic::field::{GameField};
use insta::{assert_debug_snapshot,assert_json_snapshot,assert_snapshot};
use rustyblocks::game_logic::field::render::Renderable;

#[test]
fn game_field() {
  let mut g = GameField::new(2);
  let mut scores = Vec::new();
  g.current_player = 1;
  g.current_block = 0;
  g.current_orientation = 0;
  assert_json_snapshot!("g", &g.lowest_ys);
  assert!(g.place_block_using_play(0));
  scores.push(g.scores.clone());
  assert_snapshot!("field_after_first_play", g.to_field_string());
  assert!(g.place_block_using_play(0));
  scores.push(g.scores.clone());
  assert!(g.place_block_using_play(0));
  scores.push(g.scores.clone());
  assert!(g.place_block_using_play(0));
  scores.push(g.scores.clone());
  assert_snapshot!("field_after_four_plays", g.to_field_string());
  assert_snapshot!("log_after_four_plays", serde_json::to_string(&g.log).unwrap());
  assert_debug_snapshot!("score_after_four_plays", scores);
  while !g.game_over {
    g.place_block_using_play(0);
    scores.push(g.scores.clone());
  }
  assert_snapshot!("field_after_game_over", g.to_field_string_with_player(false));
  assert_snapshot!("log_after_game_over", serde_json::to_string(&g.log).unwrap());
  assert_debug_snapshot!("scores", scores);

  g.reset(true);
  assert_snapshot!("after_reset_field", g.to_field_string());
  assert_debug_snapshot!("after_reset_scores", g.scores);
  scores.clear();
  while !g.game_over {
    g.place_block_using_play(0);
    scores.push(g.scores.clone());
  }
  assert_debug_snapshot!("after_reset_and_played_field_reshapeable", g.to_reshapeable_array(false));
  assert_debug_snapshot!("after_reset_and_played_field_reshapeable_as_second", g.to_reshapeable_array(true));
  assert_snapshot!("after_reset_and_played_field", g.to_field_string());
  assert_debug_snapshot!("after_reset_and_played_scores", g.scores);

}