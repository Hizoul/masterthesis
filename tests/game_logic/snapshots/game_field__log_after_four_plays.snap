---
created: "2019-09-18T11:15:46.641686617Z"
creator: insta@0.8.1
source: tests/game_logic/game_field.rs
expression: "serde_json::to_string(&g.log).unwrap()"
---
{"log":[{"PayloadPlaced":{"x":0,"y":0,"block":1,"orientation":0,"from":1}},{"PayloadRolled":{"from":0,"block":3}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":0,"y":2,"block":3,"orientation":0,"from":0}},{"PayloadRolled":{"from":1,"block":0}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":0,"y":4,"block":0,"orientation":0,"from":1}},{"PayloadRolled":{"from":0,"block":1}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":0,"y":5,"block":1,"orientation":0,"from":0}},{"PayloadRolled":{"from":1,"block":2}},{"PayloadConsidering":{"play_index":0}}]}
